### ISSUE TITLE
cgimap release 0.6.2

### ISSUE THREAD

**Issue by mmd-osm** [openstreetmap/operations#244]
I'm creating this follow-up operations issue, as the original pull request https://github.com/zerebubuth/openstreetmap-cgimap/pull/153 has been automatically closed during merging, making it a bit more difficult for status tracking.

A new cgimap version 0.6.2 is available for deployment.

List of changes:

  * JSON output in Overpass JSON format (https://github.com/zerebubuth/openstreetmap-cgimap/pull/153, https://github.com/zerebubuth/openstreetmap-cgimap/pull/155)
  * Include num_changes in changeset xml & json response (https://github.com/zerebubuth/openstreetmap-cgimap/pull/156)

My suggestion would be to copy the following cgimap package over to the osmadmins cgimap PPA: https://launchpad.net/~mmd-osm/+archive/ubuntu/cgimap-062/+packages - like for the 0.6.1 release.

Code comparison: https://github.com/zerebubuth/openstreetmap-cgimap/compare/v0.6.1...v0.6.2

**Comment by tomhughes** [comment]
The JSON output is enabled by default and can't be disabled right?

**Comment by mmd-osm** [comment]
Yes, the code for JSON output is enabled by default, either an Accept-Header or a .json file suffix would be needed to enabled JSON output. Both could probably also be filtered out via Rewrite rules.

**Comment by tomhughes** [comment]
It's just that our general policy has been not to have things live in cgimap that aren't support in the rails version of the code so we need to think about what to do here.

**Comment by mmd-osm** [comment]
This makes sense. I'm a bit worried though that adding JSON support to parts of a code base which aren't in productive use anymore, will be more or less wasting development resources, or most likely not being picked up at all.

Fixing JSON support in cgimap was quite a low hanging fruit, as most of the preparation work was already done by zerebubuth. I have no idea what it takes to do the same in rails.

Would it be an option to focus on a single api call, like /map and see if that could be added to rails? JSON support for others would then be marked as experimental (like a tech preview). If that's too much work, maybe declare the whole JSON support as experimental and don't bother at all with JSON support in rails?

Other approaches like cgimap-ruby might be an option at some point in the future, but I don't think that's a viable option in 2018.

**Comment by gravitystorm** [comment]
> It's just that our general policy has been not to have things live in cgimap that aren't support in the rails version of the code so we need to think about what to do here.

I'm very cautious (i.e. somewhat opposed) to enabling json support in cgimap without corresponding support in the rails port. If it can be filtered out with rewrite rules (or even better, an option added to cgimap to disable it, at least temporarily) then that would be best.

An example would be if the json support is released, then perhaps iD or JOSM will switch to it straight away. Then development of openstreetmap-website becomes much harder, since you can't get a working installation for adding data with just `rails server` any more. So you can't e.g. develop changeset comments, since you would no longer be able create changesets with only iD + openstreetmap-website. It's then some hassle for each developer to run cgimap, run another webserver like apache, configure their copy of iD within their development version of openstreetmap-website to send changes to cgimap, etc etc, just to add some data to their development database. 

It would also mean that experimental/custom/training deployments of the openstreetmap website become significantly more hassle to set up.

So I'd like to make sure that all the normal work on openstreetmap-website can continue to be developed, without making anything harder than it needs to be. But we can't realistically say "osm.org supports json but please don't use it for anything yet".

> Other approaches like cgimap-ruby might be an option at some point in the future, but I don't think that's a viable option in 2018.

The above scenarios are pretty much why we started cgimap-ruby in the first place! So I still think that it's worth focussing effort into getting this working. Then we avoid duplication of work in both directions (no need to implement every call in cgimap, no need to duplicate cgimap improvements in the rails code) and experimental/custom/training deployments finally get the benefits of cgimap 'for free'.

Is there a specific problem that means it's not viable in 2018? Or just a projection from the current rate of progress?

**Comment by tomhughes** [comment]
Making the first step "install from a PPA" immediately limits it to people running on Ubuntu...

**Comment by zerebubuth** [comment]
> An example would be if the json support is released, then perhaps iD or JOSM will switch to it straight away. Then development of openstreetmap-website becomes much harder, since you can't get a working installation for adding data with just `rails server` any more.

I'm sympathetic to the idea that Cgimap functionality should not be allowed to get ahead of the Rails server for precisely the reason you give; it makes development harder, and it's hard enough already.

However, Cgimap has had JSON support for several years without corresponding support being developed in the Rails server. Presumably, this means that no one wants the support enough to contribute it - either because it would not be useful or adding it is too hard. I suspect the latter, and if it hasn't been contributed already then it's unlikely to be contributed in the near future.

Having support only in Cgimap is a problem, but perhaps one that we can mitigate. In the short term, by having a Docker sidecar and, in the longer term, by a Cgimap-Ruby gem. (Having a development dependency on Docker isn't great, but at least it's not Ubuntu-only.) We can also ask that editor developers maintain both XML and JSON support for a while (perhaps switched by something in `/capabilities`) until JSON support is added to the Rails server.

**Comment by HolgerJeromin** [comment]
> Having a development dependency on Docker isn't great, but at least it's not Ubuntu-only.

Having a docker version gave osm-carto a bunch of new devs. So this tool does not have to be a bad thing

**Comment by bhousel** [comment]
> We can also ask that editor developers maintain both XML and JSON support for a while (perhaps switched by something in `/capabilities`) until JSON support is added to the Rails server.

This works for me 👍  I guess we could also just try using JSON optimistically and fallback to XML on error.

**Comment by tomhughes** [comment]
The problem (as you've just discovered) with the Dockerfile, and the vagrant thing, and whatever else, is that nobody uses them on a regular basis so they bitrot and whenever some poor new user makes the mistake of trying to use them they find it doesn't work any more :-(

**Comment by zerebubuth** [comment]
> The problem ... is that nobody uses them on a regular basis so they bitrot

Something that I've found can help on some projects I've worked on is to incorporate checks on the Docker/Vagrant artefact into the CI process. In other words; the CI script builds the artefact, runs some "integration" tests against it, and (if successful) pushes the image up to a public repo (perhaps only for release tags).

It's not going to catch all the possible problems, but it does make sure a basic level of functionality is maintained and that errors aren't left to compound unnoticed to the point where reviving the Docker image or whatever is a major effort.

**Comment by mmd-osm** [comment]
>  in the longer term, by a Cgimap-Ruby gem

I played a bit with the cgimap-ruby code, and tried to connect the demo web server to an apidb backend, and have it return some JSON. Although, it did in fact return some JSON after a few hours of trial and error, somehow this approach looks like quite a complex one to handle with limited ruby know-how.

https://github.com/zerebubuth/cgimap-ruby/issues/5#issuecomment-423744627 has some details.

So this confirms my prior assumption: cgimap-ruby will not fly in 2018: both the integration with the osm website code, as well as bundling it as gem is completely open at this time.

**Comment by mmd-osm** [comment]
Dockerfile is updated now, JSON enabled by default.

To build and run it, you could use something along those lines as a starting point:

<details>

```
git clone https://github.com/zerebubuth/openstreetmap-cgimap.git
cd openstreetmap-cgimap
docker build . -t cgimap
docker run --restart=always -p 8000:8000 -e "CGIMAP_HOST=localhost" -e "CGIMAP_LOGFILE=/tmp/log" -e "CGIMAP_USERNAME=osm" -e "CGIMAP_PASSWORD=pass" --net="host"  cgimap

```
</details>


lighttpd config looks a bit different now, as it points to the cgimap fastcgi server running on port 8000 inside the Docker container:


<details>

```
# default document-root
server.document-root = "/tmp/"

# TCP port
server.port = 31337

# selecting modules
server.modules = ( "mod_access", "mod_rewrite", "mod_fastcgi",  "mod_proxy", "mod_alias" )

# handling unknown routes
server.error-handler-404   = "/dispatch.map"

# read configuration from output of a command
#include_shell "/usr/local/bin/confmimetype /etc/mime.types"
mimetype.assign = (
  ".html" => "text/html",
  ".txt" => "text/plain",
  ".jpg" => "image/jpeg",
  ".png" => "image/png"
)

$HTTP["request-method"] == "GET" {
  url.rewrite-once = (
    "^/api/0\.6/map(\.(json|xml))?(\?(.*))?$" => "/dispatch.map",
    "^/api/0\.6/(node|way|relation)/[[:digit:]]+(\.(json|xml))?$" => "/dispatch.map",
    "^/api/0\.6/(node|way|relation)/[[:digit:]]+/history$" => "/dispatch.map",
    "^/api/0\.6/(node|way|relation)/[[:digit:]]+/[[:digit:]]+.*$" => "/dispatch.map",
#    "^/api/0\.6/(node|way|relation)/[[:digit:]]+/relations$" => "/dispatch.map",
#    "^/api/0\.6/node/[[:digit:]]+/ways$" => "/dispatch.map",
    "^/api/0\.6/(way|relation)/[[:digit:]]+/full$" => "/dispatch.map",
    "^/api/0\.6/changeset/[[:digit:]]+.*$" => "/dispatch.map",
    "^/api/0\.6/(nodes|ways|relations)(\?(.*))?$" => "/dispatch.map",
    "^/api/0\.6/changeset/[[:digit:]]+/download$" => "/dispatch.map",
  )
}

#$HTTP["request-method"] == "POST" {
#  url.rewrite-once = (
#    "^/api/0\.6/changeset/[[:digit:]]+/upload.*$" => "/dispatch.map",
#  )
#}

$HTTP["url"] =~ "^/(?!(dispatch\.map))" {
    proxy.server = ( "" => ( ( "host" => "127.0.0.1", "port" => "3000" ) ) )
}

fastcgi.server = ( ".map" =>
 ( "localhost" =>
   ( "host" => "127.0.0.1",
     "port" => "8000",
     "check-local" => "disable"
   )
 )
)

```

</details>

I don't know if people expect anything else here, that's about all what's needed for some osm website development uisng a cgimap Docker container as a sidecar.

**Comment by mmd-osm** [comment]
I'm seeing no further questions, concerns or objections being raised in the meantime. So I'm asking again, are we ready to deploy this production now?

**Comment by tomhughes** [comment]
Has anything changed regarding JSON being enabled?

**Comment by mmd-osm** [comment]
Not exactly, JSON is enabled by default. However, as long as Apache rewrite rules aren't adjusted yet, JSON output can only be enabled via an `HTTP-Accept: text/json` HTTP header. Appending `.json` to the URL will have no effect in this case.

**Comment by mmd-osm** [comment]
Initially, @gravitystorm raised some concerns about missing JSON support in Rails. We've discussed a number of options to mitigate this issue. Unfortunately, he hasn't commented yet, if those options are sufficient now, or if he's missing something.

Also, there's an open pull request on cgimap-ruby to include backend support, and some open questions as to how to turn it into a gem for the osm website to use. As I mentioned earlier, that's quite a bit outside of my comfort zone, and apparently not on the top prio list for @zerebubuth as well (which seems ok).

I know we're all busy and have other priorities. Still, it seems kind of difficult to find out which further steps would be needed now.

**Comment by gravitystorm** [comment]
> Not exactly, JSON is enabled by default. However, as long as Apache rewrite rules aren't adjusted yet, JSON output can only be enabled via an `HTTP-Accept: text/json` HTTP header. Appending `.json` to the URL will have no effect in this case.

Thanks for pointing this out. Unfortunately it means that [my concerns have not been addressed](https://github.com/openstreetmap/operations/issues/244#issuecomment-422330708) - we would still be releasing some form of significant new API without the corresponding API being available in our rails code. 

> I know we're all busy and have other priorities. Still, it seems kind of difficult to find out which further steps would be needed now.

Instead of the complex workarounds described above, could we have a cgimap config flag to enable json support, and have it turned off in production?

**Comment by tomhughes** [comment]
Thinking about it some more the `Accept` thing is in many ways a bigger problem than allowing a `.json` extension, as it might cause existing clients to suddenly get different results. I fear allowing JSON via `Accept` would actually need an API bump?

We could probably get apache to delete/rewrite the `Accept` header though?

**Comment by mmd-osm** [comment]
A client would have to explicitly include `text/json` in the `Accept` header. Is this something clients typically do as of today?

The code is currently deployed on the dev instance for testing, so you could try differnets `Accept` headers on an URL like `https://upload.apis.dev.openstreetmap.org/api/0.6/node/2646073` to see if this makes sense the way it is implemented.

**Comment by tomhughes** [comment]
Well we have no idea what `Accept` clients may currently be sending because they have no effect on anything.

**Comment by mmd-osm** [comment]
They do. If you send a GET request to `https://www.openstreetmap.org/api/0.6/way/166558489` with an `Accept: text/json`,  cgimap will respond with an _406 Not acceptable_ error, as an example.

**Comment by tomhughes** [comment]
Ah that's different then. I didn't realise that. I suspect the rails code doesn't, but if cgimap does then that is probably good enough to stop that being an issue.

**Comment by mmd-osm** [comment]
> we would still be releasing some form of significant new API without the corresponding API being available in our rails code.

I find this discussion a bit curious. The code on rails side we're talking about isn't in productive use  since a few years. You're basically looking at complementing existing dead code on the Rails with even more dead code - or - as we've discussed earlier, cgimap-ruby as a bridge. Unfortunately, this approach hasn't yet seen the kind of progress it needs to make it fly.

I appreciate your concerns to keep a self contained development environment around. For production use, though, I'm not convinced this is relevant in the end. It's not that we could turn off cgimap today on production, and replace with the existing Rails code without seeing some major performance impact, as an example.

Now seriously, what do we do with the cgimap-ruby gaps we need to close?

**Comment by mmd-osm** [comment]
I’d say almost all of this logic to handle ‘Accept’ is based on the implementation written by @zerebubuth about 4-8 years ago. If there are multiple equally weighted options available, xml should always win.
